src/nvim/mark_extended.c

```diff
@@ -0,0 +1,932 @@
+
+// mark_extended.c --
+// Implements extended marks for text widgets.
+// Each Mark exists in a btree of lines containing btrees
+// of columns.
+//
+// The btree provides efficent range lookus.
+// A map of pointers to the marks is used for fast lookup by mark id.
+//
+// Marks are moved by calls to extmark_col_adjust or
+// extmark_adjust which are based on col_adjust and mark_adjust from mark.c
+//
+// Undo/Redo of marks is implemented by storing the call arguments to
+// extmark_col_adjust or extmark_adjust. The list of arguments
+// is traversed in extmark_iter_undo and applied in extmark_apply_undo
+//
+// For possible ideas for efficency improvements see:
+// http://blog.atom.io/2015/06/16/optimizing-an-important-atom-primitive.html
+// Other implementations exist in gtk and tk toolkits.
+
+#include <assert.h>
+
+#include "nvim/vim.h"
+#include "nvim/mark_extended.h"
+#include "nvim/memory.h"
+#include "nvim/pos.h"          // MAXLNUM
+#include "nvim/globals.h"      // FOR_ALL_BUFFERS
+#include "nvim/map.h"          // pmap ...
+#include "nvim/lib/kbtree.h"   // kbitr ...
+#include "nvim/undo_defs.h"    // u_header_T
+#include "nvim/undo.h"         // u_save_cursor
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "mark_extended.c.generated.h"
+#endif
+
+static uint64_t namespace_counter = 0;
```

Comment:
i'm not sure i follow, `namespace_counter` is more of a hidden detail and doesn't really hinder future reuse by `bufhl`